### DISCUSSION TITLE
What is v2?

### DISCUSSION THREAD

**Discussion by ivirshup** [scikit-hep/awkward#1151] [Q&A] [answered]
I'm a little confused about the versioning of this project.

What is v2? Is there an imminent set of breaking changes? And is this described somewhere? All I could find is the roadmap which only has `1.x` releases planned.

Trying to figure out if we should be expecting things to break, and what those things might be!

**Comment by agoose77** [comment, marked as answer]
`v2` is foremostly a rewrite of the middle C++ layer of Awkward Array in pure Python. Originally, this C++ layer was intended to be usable by other C++ software (e.g. ROOT, FastJet), but it was later realised that it presented some challenges. Among these, integration with other tools such as Jax was made more difficult by the opaqueness of the C++ layer. In addition to fixing a design problem, `v2` is also being used to re-think the lazy-loading & parquet routines in Awkward. There is an [on-going project](https://iris-hep.org/projects/awkward-dask.html) to replace the `VirtualArray` and `PartitionedArray` features with a separate third-party (`dask-awkward`) library that can leverage the expertise and resources behind the Dask project to scale Awkward analysis code to clusters.

@jpivarski may correct me here, but if you're a high-level API user, that is, any public `ak.*` method, then the intent with the v1→v2 transition is that it should not hugely affect your code because the focus is updating this mid level API. In particular, this is *not* like the v0→v1 transition. Only with things like the removal of partitioned/lazy arrays (`ak.partitioned`) and removal of Parquet from core (at least, I assume we are dropping from core. This might not be the case) will you notice anything hugely disruptive. I don't know what the policy is on API breakages, but I'll leave Jim to address those!

**Reply by jpivarski** [reply]
What @agoose77 wrote is correct: a design goal of Awkward 2.0 is to avoid changes to the high-level interface, which includes the functions named `ak.this` and `ak.that`, as well as the `ak.Array`, `ak.Record`, and `ak.ArrayBuilder` classes.

Therefore, upgrading from version 1.x to version 2.0 will require minimal, if any, changes to data analyst user code. What's changing is the view that you see if you look at `.layout` of an `ak.Array`, with names like `ListOffsetArray64` and `NumpyArray`: those classes are currently C++ classes, and they're being replaced by Python classes. So, for instance, `ListOffsetArray32`, `ListOffsetArrayU32`, and `ListOffsetArray64` (three templated types in C++) are being replaced by a single `ListOffsetArray` in Python. We may provide the three old names as aliases to ease the transition, but if you make heavy use _of these layout classes_, some of the methods are changing to make the internal codebase more maintainable, and you'd have to adjust to that.

The main motivation for this refactoring ([given in detail here](https://indico.cern.ch/event/1032972/)) is to enable third-party Python libraries to be able to see more of the Awkward internals, in particular Dask. This is related to the biggest change from version 1.x to 2.0: the VirtualArray and PartitionedArray node types are being dropped in favor of an Awkward-aware Dask collection ([in development here](https://github.com/ContinuumIO/dask-awkward/), through a collaboration with Anaconda). That is, Awkward's built-in lazy arrays are being replaced by industry-standard lazy arrays. Awkward's lazy arrays were a perpetual source of trouble (see, for example, #230, #400, #432, #479, #541, #560, #597, #603, #655, #679, #783, #865, #899, #940, #1052...), largely because it evaluates the array "when some information is needed," rather than "when the user asks for it" (Dask's interface, which requires users to type `.compute()`):

<img src="https://raw.githubusercontent.com/jpivarski-talks/2021-05-21-dasksummit-awkward-collection/main/degrees-of-transparency-2.svg" width="600px">

(slide from the [Dask Summit](https://summit.dask.org/schedule/presentation/24/dask-in-high-energy-physics-community/), [slides here](https://github.com/jpivarski-talks/2021-05-21-dasksummit-awkward-collection)).

Replacing Awkward's built-in lazy arrays with Dask _is_ a big change for [Coffea's NanoEvents](https://coffeateam.github.io/coffea/notebooks/nanoevents.html), which is thoroughly virtual. The biggest part of the Awkward v1 → v2 transition will be to upgrade NanoEvents, and we'll give the Coffea team a good transition period for that work. On the plus side, it will allow Dask to be more thoroughly integrated into Coffea analyses than it currently is.

Another thing that will be different from the Awkward v0 → v1 transition is that names will not change in PyPI/conda-forge: there won't be any `awkward2` package as there was `awkward0` and `awkward1`. The downside of this is that you won't be able to load both Awkward version 1.x and 2.0 in the same Python process—the motivation for multiple names in 0.x and 1.0—but the high-level user changes would be much smaller or nonexistent in this case, so there's less motivation for being able to adopt it gradually. The name-change caused a lot more problems than I expected, so we won't be doing that again. (Also, whereas Awkward 0.x and 1.x were totally different libraries, 1.x and 2.x are more of a Ship of Theseus.)

This GitHub repo will need a new name, since scikit-hep/awkward-1.0 won't make sense anymore. scikit-hep/awkward-0.x was originally named scikit-hep/awkward-array. To be super-careful about the possibility of old links, I'm thinking the name of this repo should not become scikit-hep/awkward-array, but instead be scikit-hep/awkward, a never-before-used name. (GitHub automatically forwards links when a repo changes its name, but we don't want old links to point to the new repo, even though those old links would have to be years old. Google has already updated.)

Note that arrays pickled or otherwise saved through [ak.to_buffers](https://awkward-array.readthedocs.io/en/latest/_auto/ak.to_buffers.html)/[ak.from_buffers](https://awkward-array.readthedocs.io/en/latest/_auto/ak.from_buffers.html) (e.g. [this tutorial](https://awkward-array.org/how-to-convert-buffers.html)) will be interchangeable between versions 1.0.1+ and 2.0+. While they can't be loaded in the same Python process, they can be serialized and deserialized between the two versions (or sent through an Arrow buffer without touching disk, etc.). Awkward [version 1.0.1](https://github.com/scikit-hep/awkward-1.0/releases/tag/1.0.1), the minimum for pickle compatibility, is from Dec 14, 2020.)

I'll be talking more about the pros and cons (mostly cons) of Awkward 1.x's mixed C++/Python codebase at ACAT: [Lessons learned in Python-C++ integration](https://indico.cern.ch/event/855454/contributions/4605044/). Short story—lots of linking issues: #209, #211, #217, #233, #281, #316, #430, #476, #483, #562, #774, #778.

If you have any concerns about the performance impact of replacing C++ code with Python code, 

<img src="https://user-images.githubusercontent.com/1852447/142030610-0cb4899c-8a68-464c-a4b5-73bbb3ea7a39.png" width="600px">

(from a 2020 SciPy talk) was repeated with today's v1, v2, and the equivalent Python:

   * Awkward v1: **2.50 seconds** (yay! it's faster than in 2020, but that may either be improvements in Awkward or differences in testing, maybe order of operations and CPU cache... I don't have the original testing code)
   * Awkward v2: **1.67 seconds**: not slower in exactly the same test; in fact, it's 50% faster (more NumPy, less custom kernel)
   * Equivalent Python: **136 seconds** (almost exactly the same as in 2020, despite a new Python interpreter)

The main point of this performance test is to demonstrate that we're not putting the parts that scale with array size in pure Python: they're still in compiled code.

Also, feature requests that would have required a whole new algorithm are being implemented in version 2.0 only instead of v1 and v2 because it's easier to write once than write and port. For instance, #838 needed the `__repr__` to be rewritten to satisfy a different set of constraints, and `.tolist()` was replaced with an algorithm that's 40× faster than the one in v1. These are, technically, user-visible changes, but not part of a public API.

**Reply by ivirshup** [reply]
Thanks for the very informative responses!

We're looking at including support for awkward in [`anndata`](https://github.com/theislab/anndata) (https://github.com/theislab/anndata/pull/647), and this all sounds quite manageable!

**Comment by jpivarski** [comment]
A bit more context—the graphic below comes from a talk I'm writing.

![awkward-timeline](https://user-images.githubusercontent.com/1852447/142273956-915a48c7-2b53-4e0a-ba5a-69b38a42ff80.png)

I've also been counting lines of code and I estimate that we're 75% done with _writing_ v2. (Assuming that "lines of code" is proportional to time, that means the v2 green/development bar will continue for another 2 months.)

During the overlap period (both v1 and v2 are yellow):

   * there will be 1.x.y and 2.0.0rcZ releases in PyPI and conda-forge, both under the name `awkward` (not repeating the "name switch" mistake)
   * bugs will be fixed in 1.x.y, but new features will go into 2.0.0rcZ only (already doing that now, to some extent)
   * 2.0.0rcZ will be the main branch in GitHub and 1.x.y will be a special branch
   * the overlap period will be at least 4 months long, like v0 and v1, driven by the time needed to adapt NanoEvents and Uproot to using Dask for laziness, rather than Awkward v1 VirtualArrays

After the overlap period (v1 is gray and v2 is yellow):

   * 2.0.0 (no rc) will be released; `pip install awkward` would get that one by default (same for conda-forge)
   * the 1.x.y branch will not be deleted; bug-fixes in 1.x.y will be possible, but rare

Dropping Python 2.7 support will happen before the v1 → v2 transition, since that's driven by loss of tooling at the end of this year (#1010).

FYI @henryiii, @nsmith-, @lgray, @ioanaif, @ianna, @agoose77, @stormiestsin
